include/envoy/upstream/upstream.h

```diff
@@ -248,6 +248,12 @@ class HostSet {
    * @return uint32_t the priority of this host set.
    */
   virtual uint32_t priority() const PURE;
+
+  /**
+   * A clone of this HostSet with only hosts matching the predicate.
+   */
+  virtual std::shared_ptr<const HostSet>
+  filter(std::function<bool(const Host&)> predicate) const PURE;
```

Comment:
I think if you write this in terms of HostSharedPtr you can avoid creating wrapper lambda to derefence them when you filter.